### PR TITLE
Ensure ReferenceCountedOpenSslEngine not swallow the close_notify

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We need to ensure we not swallow the close_notify that should be send back to the remote peer. See [#6167]

Modifications:

Only call shutdown() in closeInbound() if there is nothing pending that should be send back to the remote peer.

Result:

close_notify is correctly send back to the remote peer.

### CODE REVIEW

**Comment by normanmaurer** [comment]
Also @markdascher ....

**Comment by Scottmitch** [comment]
Do we also need to modify `status` to return `SSLEngineResult.HandshakeStatus#WRAP` if there is a close_notify pending [1]?

[1] https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java#L903

**Comment by normanmaurer** [comment]
@Scottmitch it already returns `NEED_WRAP` (verified with print statements)

**Comment by Scottmitch** [comment]
@normanmaurer - is it feasible for to provide a unit test for this case?

**Comment by normanmaurer** [comment]
@Scottmitch I hope yes... gimme a day and I see if I can make it happen

**Comment by normanmaurer** [comment]
@Scottmitch the good news is I was able to write a test... that bad is that it showed another "edge-case"... Working on it as we speak ...

**Comment by normanmaurer** [comment]
@Scottmitch @markdascher PTAL again.. with unit test and everything ;)

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 1]
```diff
@@ -659,6 +662,7 @@ public final SSLEngineResult wrap(
                                 // to read more data [1]. It is also possible that event loop will detect the socket
                                 // has been closed. [1] https://www.openssl.org/docs/manmaster/ssl/SSL_write.html
                                 pendingNetResult = readPendingBytesFromBIO(dst, bytesConsumed, bytesProduced, status);
+
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: nit: revert to be consistent with other switch cases?

File: handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java [hunk 2]
```diff
@@ -1214,4 +1214,113 @@ private static void testCloseInboundAfterBeginHandshake(SSLEngine engine) throws
...
+            assertEquals(0, encryptedClientToServer.remaining());
+            assertEquals(0, plainServerOut.remaining());
+
+            assertFalse(server.isOutboundDone());
+            assertTrue(server.isInboundDone());
+
+            result = server.wrap(empty, encryptedServerToClient);
+            encryptedServerToClient.flip();
+            assertEquals(SSLEngineResult.Status.CLOSED, result.getStatus());
+            // We are done no futher UNWRAP/WRAP expected
```
Comment (Scottmitch) [inline handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java]: futher -> further

or just `UNWRAP/WRAP are not expected after this point`

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java (hunk 1 above)
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: fixed

File: handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java (hunk 2 above)
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java]: fixed

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (dd055c0) and 4.0 (bb6b251)
